File: Pattern/Modified Binary Search/240. Search a 2D Matrix II/Solution.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  let m = matrix.length;
  let n = matrix[0].length;
  let row = 0;
  let col = n - 1;
  while (row < m && col >= 0) {
    let current = matrix[row][col];
    if (current == target) {
      return true;
    } else if (current > target) {
      col--;
    } else {
      row++;
    }
  }
  return false;
};

File: pattern/Modified Binary Search/240. Search a 2D Matrix II/Solution.java
class Solution {
//  Each row is sorted in ascending order left to right

// Each column is sorted in ascending order top to bottom

// We can use a smart approach starting from the top-right corner.
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;

        int row = 0;
        int col = n - 1;

        while(row < m && col >=0){
            int current  = matrix[row][col];
            if(current == target){
                return true;
            }
            else if(current > target){
                col--;
            }
            else{
                row++;
            }
        }

        return false;
    }
}